benchmark/Benchmark.h

```diff
@@ -0,0 +1,250 @@
+// Copyright 2022, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (November of 2022, schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <vector>
+#include <functional>
+#include <string>
+#include <optional>
+
+#include "util/json.h"
+#include "util/Timer.h"
+#include <util/HashMap.h>
+#include <util/Exception.h>
+#include "../benchmark/util/HashMapWithInsertionOrder.h"
+
+/*
+ * Used for measuring the time needed for the execution of a function. It also
+ * saves the measured time togehter with a description of the function
+ * measured.
+ */
+class BenchmarkRecords {
+
+  public:
+
+    // Describes a measured function.
+    struct RecordEntry {
+      std::string descriptor_; // Needed, because without it, nobody could tell,
+                              // which time belongs to which benchmark.
+      float measuredTime_; // The measured time in seconds.
+
+    // JSON (d)serialization for all the created structures in BenchmarkRecords.
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(BenchmarkRecords::RecordEntry, descriptor_, measuredTime_)
+    };
+
+    // Describes a group of measured functions.
+    struct RecordGroup {
+      std::string descriptor_; // Needed for identifying groups.
+      std::vector<RecordEntry> entries_;
+
+    // JSON (d)serialization for all the created structures in BenchmarkRecords.
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(BenchmarkRecords::RecordGroup, descriptor_, entries_)
+    };
+    
+    // Describes a table of measured functions.
+    struct RecordTable {
+      // For identification.
+      std::string descriptor_;
+      // The names of the columns and rows.
+      std::vector<std::string> rowNames_;
+      std::vector<std::string> columnNames_;
+      // The time measurements in seconds. Access is [row, column].
+      std::vector<std::vector<std::optional<float>>> entries_;
+
+      RecordTable(const std::string& pDescriptor, 
+          const std::vector<std::string>& pRowNames,
+          const std::vector<std::string>& pColumnNames):
+          descriptor_{pDescriptor}, rowNames_{pRowNames},
+          columnNames_{pColumnNames},
+          entries_(pRowNames.size(), std::vector<std::optional<float>>(pColumnNames.size())) {}
+
+      // JSON (d)serialization for all the created structures in BenchmarkRecords.
+      NLOHMANN_DEFINE_TYPE_INTRUSIVE(BenchmarkRecords::RecordTable, descriptor_, rowNames_, columnNames_, entries_)
+    };
```

Comment:
Add support for metadata, so that information about benchmarks can be saved for later ordering and searching of them.

We don't have a metadata system, so you are going to have to make it yourself.

Reply:
Wrote first draft in [da58ed3](https://github.com/ad-freiburg/qlever/pull/860/commits/da58ed3bbd4ea7540796077fac6e168fff2d82ef) and finished the implementation in [e6e570f](https://github.com/ad-freiburg/qlever/pull/860/commits/e6e570f0bcecb4d7bf89bfc4ca2bf29b7d92f68f).

P.S.: I have a slight concern about the serialized form of the metadata class. (See the concerning TODO in the class.) What do you think, @joka921 ? Is it alright, or should it be changed?

Reply:
Seems to be alright.